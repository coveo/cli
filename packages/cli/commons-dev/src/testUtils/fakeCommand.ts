import {Command} from '@oclif/core';

export const getFakeCommand = (
  overrideConfig?: Record<string, unknown>
): Command => {
  const fakeCommand = {
    id: 'foo',
    warn: jest.fn(),
    log: jest.fn(),
    parse: jest.fn().mockReturnValue({flags: {}}),
    ...overrideConfig,
  };

  return fakeCommand as unknown as Command;
};
