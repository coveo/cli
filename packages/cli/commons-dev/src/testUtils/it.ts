import {fancy} from 'fancy-test';

/**
 * Wrapper around `jest.it` that captures all stdout/stderr.
 */
export const fancyIt = () => fancy.stdout().stderr().it;
