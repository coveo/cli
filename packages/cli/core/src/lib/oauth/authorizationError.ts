import dedent from 'ts-dedent';
import type {URL} from 'url';
import {CLIBaseError} from '@coveo/cli-commons/errors/cliBaseError';

export class AuthorizationError extends CLIBaseError {
  public name = 'Authorization Error';
  public constructor(public url: URL) {
    super();

    this.message = [
      url.searchParams.get('error'),
      url.searchParams.get('error_description'),
      url.searchParams.get('error_uri'),
    ]
      .filter((value) => value !== null)
      .join('\n');
  }
}

export class InvalidStateError extends CLIBaseError {
  public name = 'Invalid State Error';
  public constructor(receivedState: string, expectedState: string) {
    super();

    this.message = dedent`Unexpected state in authorization request.
    Expected state: ${expectedState}
    Received state: ${receivedState}`;
  }
}

export class OAuthPortBusy extends CLIBaseError {
  public name = 'OAuth ports busy';
  public constructor() {
    super();

    this.message = dedent`
      TCP port 32111 and 52296 are both busy.
      The CLI needs at least of one of them to complete the login process.
      Free one of them or use \`coveo auth:token\` to authenticate.
    `;
  }
}
