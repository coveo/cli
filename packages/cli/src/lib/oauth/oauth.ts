import {
  DEFAULT_ENVIRONMENT,
  DEFAULT_REGION,
  PlatformEnvironment,
  platformUrl,
} from '../platform/environment';
import {
  AuthorizationNotifier,
  AuthorizationRequest,
  AuthorizationServiceConfiguration,
  AuthorizationServiceConfigurationJson,
  BaseTokenRequestHandler,
  GRANT_TYPE_AUTHORIZATION_CODE,
  TokenRequest,
} from '@openid/appauth';
import {
  NodeBasedHandler,
  NodeCrypto,
  NodeRequestor,
} from '@openid/appauth/built/node_support';
import {Region} from '@coveord/platform-client';

export interface OAuthOptions {
  port: number;
  environment: PlatformEnvironment;
  region: Region;
}

export class OAuth {
  private opts: OAuthOptions;
  public constructor(opts?: Partial<OAuthOptions>) {
    const baseOptions: OAuthOptions = {
      port: 32111,
      environment: DEFAULT_ENVIRONMENT,
      region: DEFAULT_REGION,
    };

    this.opts = {
      environment: opts?.environment || baseOptions.environment,
      port: opts?.port || baseOptions.port,
      region: opts?.region || baseOptions.region,
    };
  }

  public async getToken() {
    const config = new AuthorizationServiceConfiguration({
      ...this.clientConfig,
      ...this.authServiceConfig,
    });

    const {code} = await this.getAuthorizationCode(config);
    return await this.getAccessToken(config, code);
  }

  private async getAccessToken(
    configuration: AuthorizationServiceConfiguration,
    code: string
  ): Promise<{accessToken: string}> {
    const tokenHandler = new BaseTokenRequestHandler(new NodeRequestor());

    const request = new TokenRequest({
      ...this.clientConfig,
      grant_type: GRANT_TYPE_AUTHORIZATION_CODE,
      code,
      extras: {
        client_secret: this.clientConfig.client_id,
      },
    });

    try {
      const response = await tokenHandler.performTokenRequest(
        configuration,
        request
      );
      return response;
    } catch (e) {
      console.error('ERROR: ', e as string);
      throw e;
    }
  }

  private getAuthorizationCode(
    configuration: AuthorizationServiceConfiguration
  ): Promise<{code: string; verifier: string}> {
    return new Promise((res) => {
      const notifier = new AuthorizationNotifier();
      const authorizationHandler = new NodeBasedHandler(this.opts.port);
      const request = new AuthorizationRequest(
        {
          ...this.clientConfig,
          response_type: AuthorizationRequest.RESPONSE_TYPE_CODE,
        },
        new NodeCrypto(),
        true
      );

      authorizationHandler.setAuthorizationNotifier(notifier);
      notifier.setAuthorizationListener((request, response) => {
        if (response) {
          const {code} = response;
          res({code, verifier: request.internal!.verifier});
        }
      });
      authorizationHandler.performAuthorizationRequest(configuration, request);
    });
  }

  private get clientConfig() {
    return {
      client_id: 'cli',
      redirect_uri: `http://127.0.0.1:${this.opts.port}`,
      scope: 'full',
    };
  }

  private get authServiceConfig(): AuthorizationServiceConfigurationJson {
    const baseURL = platformUrl({
      environment: this.opts.environment,
      region: this.opts.region,
    });
    return {
      authorization_endpoint: `${baseURL}/oauth/authorize`,
      revocation_endpoint: `${baseURL}/logout`,
      token_endpoint: `${baseURL}/oauth/token`,
    };
  }
}
