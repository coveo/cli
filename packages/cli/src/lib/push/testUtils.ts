export const doMockAxiosError = (
  status: number,
  message: string,
  errorCode: string
) => ({
  response: {
    status,
    data: {
      errorCode,
      message,
    },
  },
});

export const doMockAxiosSuccess = (status: number, text: string) => ({
  status,
  statusText: text,
});
