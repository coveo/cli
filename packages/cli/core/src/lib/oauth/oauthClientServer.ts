import 'fetch-undici-polyfill';
import {createServer, IncomingMessage, Server, ServerResponse} from 'http';
import {URL, URLSearchParams} from 'url';
import {AuthorizationError, InvalidStateError} from './authorizationError';
import {AuthorizationServiceConfiguration, ClientConfig} from './oauthConfig';
interface AccessTokenResponse {
  access_token: string;
}

export class OAuthClientServer {
  public constructor(
    private clientConfig: ClientConfig,
    private authServiceConfig: AuthorizationServiceConfiguration
  ) {}

  public startServer(port: number, hostname: string, expectedState: string) {
    let server: Server;
    const serverPromise = new Promise<{accessToken: string}>(
      (resolve, reject) => {
        const listener = async (req: IncomingMessage, res: ServerResponse) => {
          try {
            const {code, state} = this.parseUrl(req);
            const {tokenEndpoint} = this.authServiceConfig;

            if (!state || !code) {
              return;
            }

            if (state !== expectedState) {
              throw new InvalidStateError(state, expectedState);
            }
            const body = this.getTokenQueryString(code);
            const authRequest = await fetch(tokenEndpoint, {
              method: 'POST',
              body,
              headers: this.requestHeaders,
            });
            const {access_token: accessToken} =
              (await authRequest.json()) as AccessTokenResponse;

            res.end('Close your browser to continue');
            resolve({accessToken});
          } catch (error: unknown) {
            reject(error);
          } finally {
            server.close();
          }
        };

        server = createServer(listener);
        server.listen(port, hostname);
      }
    );

    return serverPromise;
  }

  private get requestHeaders(): Record<string, string> {
    const {client_id} = this.clientConfig;

    return {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization: `Basic ${Buffer.from(`${client_id}:${client_id}`).toString(
        'base64'
      )}`,
    };
  }

  private parseUrl(req: IncomingMessage) {
    if (!req.url) {
      return {
        state: null,
        code: null,
      };
    }

    const url = new URL(`http://127.0.0.1${req.url}`);
    const state = url.searchParams.get('state');
    const code = url.searchParams.get('code');
    const error = url.searchParams.get('error');

    if (error) {
      throw new AuthorizationError(url);
    }

    return {state, code};
  }

  private getTokenQueryString(code: string) {
    const {redirect_uri} = this.clientConfig;
    const params = {
      grant_type: 'authorization_code',
      redirect_uri,
      code,
    };

    return new URLSearchParams(params);
  }
}
