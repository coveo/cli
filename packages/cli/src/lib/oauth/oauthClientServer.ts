import axios, {AxiosRequestConfig} from 'axios';
import {createServer, IncomingMessage, Server, ServerResponse} from 'http';
import {URLSearchParams} from 'url';
import {AuthorizationError, InvalidStateError} from './authorizationError';
import {AuthorizationServiceConfiguration, ClientConfig} from './oauthConfig';

interface AccessTokenResponse {
  access_token: string;
}

export class OAuthClientServer {
  public constructor(
    private clientConfig: ClientConfig,
    private authServiceConfig: AuthorizationServiceConfiguration
  ) {}

  public startServer(port: number, hostname: string, expectedState: string) {
    let server: Server;
    const serverPromise = new Promise<{accessToken: string}>(
      (resolve, reject) => {
        const listener = async (req: IncomingMessage, res: ServerResponse) => {
          try {
            const {code, state} = this.parseUrl(req);
            const {tokenEndpoint} = this.authServiceConfig;

            if (!state || !code) {
              return;
            }

            if (state !== expectedState) {
              throw new InvalidStateError(state, expectedState);
            }

            const data = this.getTokenQueryString(code);
            const authRequest = await axios.post<AccessTokenResponse>(
              tokenEndpoint,
              data,
              this.requestConfig
            );
            const accessToken = authRequest.data.access_token;

            res.end('Close your browser to continue');
            resolve({accessToken});
          } catch (error: unknown) {
            reject(error);
          } finally {
            server.close();
          }
        };

        server = createServer(listener);
        server.listen(port, hostname);
      }
    );

    return serverPromise;
  }

  private get requestConfig(): AxiosRequestConfig {
    const {client_id} = this.clientConfig;
    const config: AxiosRequestConfig = {
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      auth: {
        username: client_id,
        password: client_id,
      },
    };

    return config;
  }

  private parseUrl(req: IncomingMessage) {
    if (!req.url) {
      return {
        state: null,
        code: null,
      };
    }

    const url = new URL(`http://127.0.0.1${req.url}`);
    const state = url.searchParams.get('state');
    const code = url.searchParams.get('code');
    const error = url.searchParams.get('error');

    if (error) {
      throw new AuthorizationError(url);
    }

    return {state, code};
  }

  private getTokenQueryString(code: string) {
    const {redirect_uri} = this.clientConfig;
    const params = {
      grant_type: 'authorization_code',
      redirect_uri,
      code,
    };

    return new URLSearchParams(params);
  }
}
