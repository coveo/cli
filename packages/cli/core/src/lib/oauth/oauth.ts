import open from 'open';
import {
  DEFAULT_ENVIRONMENT,
  DEFAULT_REGION,
  PlatformEnvironment,
  platformUrl,
} from '@coveo/cli-commons/platform/environment';
import {Region} from '@coveo/platform-client';
import {randomBytes} from 'crypto';
import {AuthorizationServiceConfiguration, ClientConfig} from './oauthConfig';
import {OAuthClientServer} from './oauthClientServer';
import {URL} from 'url';
import getPort from 'get-port';
import {OAuthPortBusy} from './authorizationError';

interface OAuthOptions {
  environment: PlatformEnvironment;
  region: Region;
}

export class OAuth {
  private static readonly AllowedPorts = [32111, 52296];

  private opts: OAuthOptions;
  private _port: number | undefined = undefined;

  public constructor(opts?: Partial<OAuthOptions>) {
    const baseOptions: OAuthOptions = {
      environment: DEFAULT_ENVIRONMENT,
      region: DEFAULT_REGION,
    };

    this.opts = {
      environment: opts?.environment || baseOptions.environment,
      region: opts?.region || baseOptions.region,
    };
  }

  public async getToken() {
    const state = this.generateState(10);
    const requestHandler = new OAuthClientServer(
      await this.getClientConfig(),
      this.authServiceConfig
    );

    const clientServerPromise = requestHandler.startServer(
      await this.getPort(),
      '127.0.0.1',
      state
    );

    await this.openLoginPage(state);

    return clientServerPromise;
  }

  private async openLoginPage(state: string) {
    const url = await this.buildAuthorizationUrl(state);
    open(url).then((browserProcess) => {
      browserProcess.unref();
    });
  }

  private async buildAuthorizationUrl(state: string) {
    const {authorizationEndpoint} = this.authServiceConfig;
    const {redirect_uri, client_id, scope} = await this.getClientConfig();
    const url = new URL(authorizationEndpoint);

    url.searchParams.append('response_type', 'code');
    url.searchParams.append('client_id', client_id);
    url.searchParams.append('redirect_uri', redirect_uri);
    url.searchParams.append('scope', scope);
    url.searchParams.append('state', state);

    return url.href;
  }

  private generateState(size: number) {
    return randomBytes(size).toString('hex');
  }

  private async getClientConfig(): Promise<ClientConfig> {
    return {
      client_id: 'cli',
      redirect_uri: `http://127.0.0.1:${await this.getPort()}`,
      scope: 'full',
    };
  }

  private get authServiceConfig(): AuthorizationServiceConfiguration {
    const baseURL = platformUrl({
      environment: this.opts.environment,
      region: this.opts.region,
    });
    return {
      authorizationEndpoint: `${baseURL}/oauth/authorize`,
      revocationEndpoint: `${baseURL}/logout`,
      tokenEndpoint: `${baseURL}/oauth/token`,
    };
  }

  private async getPort(): Promise<number> {
    if (!this._port) {
      const candidate = await getPort({port: OAuth.AllowedPorts});
      if (!OAuth.AllowedPorts.includes(candidate)) {
        throw new OAuthPortBusy();
      }
      this._port = candidate;
    }
    return this._port;
  }
}
