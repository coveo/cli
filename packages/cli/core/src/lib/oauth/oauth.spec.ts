import type {ChildProcess} from 'child_process';

import {Region} from '@coveord/platform-client';
import {fancyIt} from '@coveo/cli-commons-dev/testUtils/it';
import {
  PlatformEnvironment,
  platformUrl,
} from '@coveo/cli-commons/platform/environment';
import {OAuth} from './oauth';
import {OAuthClientServer} from './oauthClientServer';
import open from 'open';

jest.mock('./oauthClientServer');
jest.mock('open');

describe('OAuth', () => {
  const mockedOauthClientServer = jest.mocked(OAuthClientServer);
  const mockedOpen = jest.mocked(open);
  const mockedStartServer = jest.fn();

  beforeAll(() => {
    mockedOauthClientServer.prototype.startServer.mockImplementation(
      mockedStartServer
    );
    mockedOpen.mockResolvedValue({unref: () => {}} as ChildProcess);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  fancyIt()('should return the access token when requested', async () => {
    mockedStartServer.mockResolvedValueOnce({
      accessToken: 'this-is-the-new-access-token',
    });
    const {accessToken} = await new OAuth().getToken();
    expect(accessToken).toEqual('this-is-the-new-access-token');
  });

  describe('should use proper port for client server', () => {
    fancyIt()('using default port', async () => {
      await new OAuth().getToken();
      expect(mockedStartServer).toHaveBeenCalledWith(
        32111,
        expect.anything(),
        expect.anything()
      );
    });

    fancyIt()('using configured port', async () => {
      await new OAuth({port: 444}).getToken();
      expect(mockedStartServer).toHaveBeenCalledWith(
        444,
        expect.anything(),
        expect.anything()
      );
    });
  });

  describe('should use proper ClientConfig', () => {
    const endpoints = (opts?: {
      environment?: PlatformEnvironment;
      region?: Region;
    }) => ({
      authorizationEndpoint: `${platformUrl(opts)}/oauth/authorize`,
      revocationEndpoint: `${platformUrl(opts)}/logout`,
      tokenEndpoint: `${platformUrl(opts)}/oauth/token`,
    });

    const defaultClientConfig = {
      client_id: 'cli',
      redirect_uri: 'http://127.0.0.1:32111',
      scope: 'full',
    };

    fancyIt()('in prod', async () => {
      const opts = {environment: PlatformEnvironment.Prod};
      await new OAuth(opts).getToken();
      expect(mockedOauthClientServer).toHaveBeenCalledWith(
        defaultClientConfig,
        endpoints(opts)
      );
    });

    fancyIt()('in stg', async () => {
      const opts = {environment: PlatformEnvironment.Stg};
      await new OAuth(opts).getToken();
      expect(mockedOauthClientServer).toHaveBeenCalledWith(
        defaultClientConfig,
        endpoints(opts)
      );
    });

    fancyIt()('in dev', async () => {
      const opts = {environment: PlatformEnvironment.Dev};
      await new OAuth(opts).getToken();
      expect(mockedOauthClientServer).toHaveBeenCalledWith(
        defaultClientConfig,
        endpoints(opts)
      );
    });

    fancyIt()('in hipaa', async () => {
      const opts = {environment: PlatformEnvironment.Stg};
      await new OAuth(opts).getToken();
      expect(mockedOauthClientServer).toHaveBeenCalledWith(
        defaultClientConfig,
        endpoints(opts)
      );
    });

    fancyIt()('in Europe', async () => {
      const opts = {region: Region.EU};
      await new OAuth(opts).getToken();
      expect(mockedOauthClientServer).toHaveBeenCalledWith(
        defaultClientConfig,
        endpoints(opts)
      );
    });

    fancyIt()('it should be U.S. prod by default', async () => {
      await new OAuth().getToken();
      expect(mockedOauthClientServer).toHaveBeenCalledWith(
        defaultClientConfig,
        endpoints({
          environment: PlatformEnvironment.Prod,
          region: Region.US,
        })
      );
    });
  });
});
