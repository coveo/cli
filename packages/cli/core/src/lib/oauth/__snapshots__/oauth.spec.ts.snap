// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuth should throw an OAuthBusyError if both OAuth ports are busy 1`] = `
"TCP port 32111 and 52296 are both busy.
The CLI needs at least of one of them to complete the login process.
Free one of them or use \`coveo auth:token\` to authenticate."
`;
